pkg/credentials/buffer/ring_buffer_test.go

```diff
@@ -0,0 +1,376 @@
+			}
+		})
+	}
+}
+
+func TestNotIsEmptyAfterPoppingPartially(t *testing.T) {
+	buf := buffer.NewRingBuffer(3)
+
+	buf.Add(1)
+	buf.Add(2)
+	buf.Add(3)
+
+	_, err := buf.Pop(2)
+	if err != nil {
+		t.Fatal(err)
+	}
+	empty := buf.IsEmpty()
+	if empty {
+		t.Error("Expected buffer to be not empty")
+	}
+}
+
+func TestIsEmptyAfterPoppingAllBytes(t *testing.T) {
+	buf := buffer.NewRingBuffer(2)
+
+	buf.Add(1)
+	buf.Add(2)
+
+	_, err := buf.PopFirst()
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = buf.PopFirst()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	isEmpty := buf.IsEmpty()
+	if !isEmpty {
+		t.Error("expected buffer to be empty at end")
+	}
+}
+
+func TestGet(t *testing.T) {
+	type getTest struct {
+		capacity        int
+		bytesToAdd      []byte
+		indexToRetrieve int
+		expectedByte    byte
+		expectedError   bool
+	}
+	testCases := []getTest{
+		{0, []byte{}, -1, 0, true},
+		{0, []byte{}, 0, 0, true},
+		{1, []byte{}, 1, 0, true},
+		{1, []byte{}, -1, 0, true},
+		{3, []byte{1, 2}, 2, 0, true},
+		{3, []byte{1, 2, 3}, 0, 1, false},
+		{3, []byte{1, 2, 3}, 1, 2, false},
+		{3, []byte{1, 2, 3}, 2, 3, false},
+	}
+
+	for _, tc := range testCases {
+		name := fmt.Sprintf("cap_%d_index_%d_byte_%b_err_%t_withbytes_%v", tc.capacity,
+			tc.indexToRetrieve, tc.expectedByte, tc.expectedError, tc.bytesToAdd)
+		t.Run(name, func(t1 *testing.T) {
+			buf := buffer.NewRingBuffer(tc.capacity)
+
+			buf.AddMulti(tc.bytesToAdd)
+
+			b, err := buf.Get(tc.indexToRetrieve)
+			if tc.expectedError && err == nil {
+				t1.Error("expected error but got nil")
+			} else if !tc.expectedError && err != nil {
+				t1.Error(err)
+			} else if tc.expectedByte != b {
+				t1.Errorf("expected byte %b but got %b", tc.expectedByte, b)
+			}
+		})
+	}
+}
+
+func TestGetDoesNotChangeBuffer(t *testing.T) {
+	buf := buffer.NewRingBuffer(5)
+	buf.Add(1)
+	buf.Add(2)
+	buf.Add(3)
+
+	// get twice to check that it is not popped/modified
+	_, err := buf.Get(0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	one, err := buf.Get(0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = buf.Get(1)
+	if err != nil {
+		t.Fatal(err)
+	}
+	two, err := buf.Get(1)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = buf.Get(2)
+	if err != nil {
+		t.Fatal(err)
+	}
+	three, err := buf.Get(2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	empty := buf.IsEmpty()
+	if empty {
+		t.Error("Expected buffer to be not empty")
+	}
+	if one != 1 {
+		t.Errorf("First byte should be 1, was %b", one)
+	}
+	if two != 2 {
+		t.Errorf("Second byte should be 2, was %b", two)
+	}
+	if three != 3 {
+		t.Errorf("Third byte should be 3, was %b", three)
+	}
+}
+
+func TestPopFirst(t *testing.T) {
+	type popFirstTest struct {
+		capacity      int
+		bytesToAdd    []byte
+		expectedBytes []byte
+		expectedError bool
+	}
+	testCases := []popFirstTest{
+		{0, []byte{}, []byte{}, false},
+		{0, []byte{}, []byte{0}, true},
+		{3, []byte{}, []byte{0}, true},
+		{3, []byte{}, []byte{0}, true},
```

Comment:
This is probably just confusing. The first test case pops 0 bytes from the buffer (length of the second slice) and that is fine because the buffer can in principal serve that, even if it is an unusal case.
The second test case pops 1 byte from the buffer which has zero capacity. That is clearly an error.

Reply:
I added a comment.

Reply:
Thanks, that's helpful. It looks like the 3rd and 4th test cases are identical though (here and in the next test class).